eng/packages/http-client-csharp/generator/Azure.Generator/src/Visitors/LroVisitor.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Azure.Generator.Primitives;
+using Azure.Generator.Providers;
+using Microsoft.TypeSpec.Generator.ClientModel;
+using Microsoft.TypeSpec.Generator.ClientModel.Providers;
+using Microsoft.TypeSpec.Generator.Expressions;
+using Microsoft.TypeSpec.Generator.Primitives;
+using Microsoft.TypeSpec.Generator.Providers;
+using static Microsoft.TypeSpec.Generator.Snippets.Snippet;
+
+namespace Azure.Generator.Visitors
+{
+    internal class LroVisitor : ScmLibraryVisitor
+    {
+        protected override ScmMethodProvider? VisitMethod(ScmMethodProvider method)
+        {
+            if (method.IsLroMethod())
+            {
+                var responseType = method.ServiceMethod!.Response.Type;
+
+                // No convenience method is needed for an LRO method that doesn't return a response type
+                if (responseType == null && !method.IsProtocolMethod)
+                {
+                    return null;
+                }
+
+                var returnType = (responseType, method.IsProtocolMethod) switch
+                {
+                    (null, true) => typeof(Operation),
+                    (not null, true) => new CSharpType(typeof(Operation), typeof(BinaryData)),
+                    _ => new CSharpType(typeof(Operation<>), AzureClientGenerator.Instance.TypeFactory.CreateCSharpType(responseType!)!),
+                };
+                var isAsync = method.Signature.Modifiers.HasFlag(MethodSignatureModifiers.Async);
+
+                // Update the method signature
+                var parameters = new List<ParameterProvider>(method.Signature.Parameters);
+                parameters.Insert(0, new ParameterProvider(
+                    name: "waitUntil",
+                    type: typeof(WaitUntil),
+                    description: FormattableStringFactory.Create(
+                        "<see cref=\"WaitUntil.Completed\"/> if the method should wait to return until the long-running operation " +
+                        "has completed on the service; <see cref=\"WaitUntil.Started\"/> if it should return after starting the operation. " +
+                        "For more information on long-running operations, please see " +
+                        "<see href=\"https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/LongRunningOperations.md\"> " +
+                        "Azure.Core Long-Running Operation samples</see>.")));
+
+                if (method.IsProtocolMethod && responseType == null)
+                {
+                    // Make the requestContext parameter optional as there is no corresponding convenience method to worry about ambiguous calls
```

Comment:
does this make it more difficult if later the library wants to add a convenience method?

Reply:
No, we do not generate convenience methods for LROs that don't have a response/request payload as the protocol method has all of the same parameters except for the RequestContext vs CancellationToken. But thinking about this more, this isn't really specific to LROs and we don't have the same handling for regular operations. I think we should probably be consistent and just generate both the protocol and the convenience method here even though it is different than the behavior in autorest.